src/internal/dbutil/tx.go

```diff
@@ -146,16 +145,62 @@ func WithTx(ctx context.Context, db *pachsql.DB, cb func(cbCtx context.Context,
 	}()
 
 	txStartedMetric.Inc()
-	err := backoff.RetryUntilCancel(ctx, func() error {
-		ctx, cf := pctx.WithCancel(pctx.Child(ctx, fmt.Sprintf("WithTx(%d)", attempts)))
+	err := backoff.RetryUntilCancel(ctx, func() (retErr error) {
+		attemptStart := time.Now()
+		ctx, cf := pctx.WithCancel(ctx)
 		defer cf()
+
+		ctx, done := log.SpanContext(ctx, "WithTx", zap.Int("tx_attempt", attempts))
+		var doneFields []log.Field
+		defer func() {
+			if !errors.Is(retErr, sql.ErrTxDone) {
+				// It's a common pattern that we call tx.Rollback ourselves, and
+				// then the tryTxFunc calls tx.Commit and gets an error.  That
+				// should not fail the span, so only add a zap.Error field when the
+				// error is not that type.
+				doneFields = append(doneFields, zap.Error(retErr))
+			}
+			done(doneFields...)
+		}()
+
 		underlyingTxStartedMetric.Inc()
 		attempts++
 		tx, err := db.BeginTxx(ctx, &c.TxOptions)
 		if err != nil {
 			underlyingTxFinishMetric.WithLabelValues("failed_start").Inc()
 			return errors.EnsureStack(err)
 		}
+
+		// Note: query logging includes a pgx.pid field that does not match this one.
+		// That's because it's the pid on the pgbouncer side (a property that is negotiated
+		// as part of the connection setup), whereas the one we're fetching here is the pid
+		// on the postgres side.  This one is what you can look up in pg_stat_activity, so
+		// is more useful.
+		var xactId, pid uint32
+		row := tx.QueryRowxContext(ctx, "select pg_current_xact_id(), pg_backend_pid()")
+		if err := row.Scan(&xactId, &pid); err != nil {
+			log.Debug(ctx, "failed to fetch current transaction and backend ids", zap.Error(err))
+		} else {
+			// If the query worked, then include xact_id and pid on tryTxFunc's logs
+			// (which will often be pgx's logs when query logging is enabled).  We add
+			// them to doneFields so that if tryTxFunc doesn't log anything, we still
+			// get them when the span is complete.
+			doneFields = append(doneFields, zap.Uint32("xact_id", xactId), zap.Uint32("pid", pid))
+			ctx = pctx.Child(ctx, "", pctx.WithFields(doneFields...))
+		}
+
+		// Report long transactions.
+		go func() {
+			for {
```

Comment:
would this have the potential to flood the logs with a ton of entries? i'm not seeing any sort of delay between loop iterations but maybe that is just me being ignorant of go's semantics

Reply:
<-time.After is like a sleep statement.  So it returns when the context is done (which is when the containing function returns because of the pctx.WithCancel at the top), or every 10 seconds.